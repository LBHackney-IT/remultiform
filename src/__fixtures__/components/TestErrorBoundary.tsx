import PropTypes from "prop-types";
import React from "react";

export interface TestErrorBoundaryProps {
  children: React.ReactNode;
}

export interface TestErrorBoundaryState {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  error?: any;
}

export class TestErrorBoundary extends React.Component<
  TestErrorBoundaryProps,
  TestErrorBoundaryState
> {
  static propTypes: PropTypes.ValidationMap<TestErrorBoundaryProps> = {
    children: PropTypes.node.isRequired,
  };

  static getDerivedStateFromError(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    error: any
  ): Partial<TestErrorBoundaryState> {
    return { error };
  }

  state: TestErrorBoundaryState = {};

  render(): JSX.Element {
    const { children } = this.props;
    const { error } = this.state;

    if (error) {
      return <span>Caught error: {error.toString()}</span>;
    }

    return <>{children}</>;
  }
}
