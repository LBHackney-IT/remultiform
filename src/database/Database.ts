import {
  IDBPDatabase,
  StoreKey as IDBPStoreKey,
  StoreValue as IDBPStoreValue
} from "idb";

import { wrapOpenDB } from "./internal/wrapOpenDB";
import { wrapTransaction } from "./internal/wrapTransaction";

import { OpenOptions } from "./OpenOptions";
import { StoreMap } from "./Store";
import {
  NamedSchema,
  Schema,
  StoreKey,
  StoreNames,
  StoreValue,
  TransactionMode
} from "./types";

/**
 * A wrapper for an IndexedDB connection.
 *
 * Call {@link Database.open} to create a new database connection.
 *
 * @typeparam NamedDBSchema - The schema for the database, along with the
 * allowed names for databases using that schema.
 */
export class Database<DBSchema extends NamedSchema<string, number, Schema>> {
  /**
   * Open a new database connection.
   *
   * @typeparam NamedDBSchema - The schema for the database, along with the
   * allowed names for databases using that schema.
   *
   * @param name - The name of the database.
   *
   * @param version - The schema version you want to open. This can only ever
   * increase. See {@link OpenOptions.upgrade} for how to upgrade the schema
   * when changing version.
   *
   * @param openOptions - The options and callbacks for opening the database.
   * See {@link OpenOptions} for details.
   *
   * @returns A promise to an open {@link Database} ready to be accessed.
   */
  static async open<DBSchema extends NamedSchema<string, number, Schema>>(
    name: DBSchema["dbNames"],
    version: DBSchema["versions"],
    openOptions: OpenOptions<DBSchema["schema"]> = {}
  ): Promise<Database<DBSchema>> {
    const db = await wrapOpenDB(name, version, openOptions);

    return new Database(db);
  }

  /**
   * The name of the IndexedDB database.
   */
  readonly name: string;

  /**
   * The `idb` wrapper of the IndexedDB database this {@link Database} wraps.
   */
  private readonly db: IDBPDatabase<DBSchema["schema"]>;

  /**
   * Do not use this directly. Use {@link Database.open} to create a new
   * {@link Database}.
   *
   * @ignore
   */
  constructor(db: IDBPDatabase<DBSchema["schema"]>) {
    this.db = db;

    this.name = this.db.name;
  }

  /**
   * Create or update a value in a store.
   */
  async put<DBStoreName extends StoreNames<DBSchema["schema"]>>(
    storeName: DBStoreName,
    key: StoreKey<DBSchema["schema"], DBStoreName>,
    value: StoreValue<DBSchema["schema"], DBStoreName>
  ): Promise<void> {
    await this.db.put(
      storeName,
      (value as unknown) as IDBPStoreValue<DBSchema["schema"], DBStoreName>,
      key as IDBPStoreKey<DBSchema["schema"], DBStoreName>
    );
  }

  /**
   * Get a value from a store.
   *
   * Resolves to `undefined` if no value exists for the given key in the given
   * store.
   */
  get<DBStoreName extends StoreNames<DBSchema["schema"]>>(
    storeName: DBStoreName,
    key: StoreKey<DBSchema["schema"], DBStoreName>
  ): Promise<StoreValue<DBSchema["schema"], DBStoreName> | undefined> {
    return this.db.get(
      storeName,
      key as IDBPStoreKey<DBSchema["schema"], DBStoreName>
    );
  }

  /**
   * Create a transaction on one or more stores.
   *
   * @param storeNames - The stores to open the transaction over.
   *
   * @param tx - A callback for the transaction to call. Be aware that this
   * callback, although possibly asynchronous, needs to be updating the
   * transaction every tick. If no changes to the transaction happen in a tick,
   * it will automatically commit itself and the transaction will close.
   *
   * @param mode - The mode to open the transaction in.
   */
  async transaction<DBStoreNames extends StoreNames<DBSchema["schema"]>[]>(
    storeNames: DBStoreNames,
    tx: (
      stores: StoreMap<DBSchema["schema"], DBStoreNames>
    ) => void | Promise<void>,
    mode: TransactionMode = TransactionMode.ReadOnly
  ): Promise<void> {
    const transaction = this.db.transaction(storeNames, mode);

    await wrapTransaction(storeNames, transaction, tx);
  }

  /**
   * Request to close the database connection.
   *
   * If there are no open transactions, this will close immediately. If any
   * transactions are open, it will wait for them to complete before closing.
   * This happens asynchronously, out of band.
   */
  // It would be better if this waited for the connection to close, but
  // `fake-indexeddb` doesn't support the `close` event we need, making it
  // difficult to implement and difficult to test.
  close(): void {
    this.db.close();
  }
}
