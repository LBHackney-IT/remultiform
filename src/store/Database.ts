import { IDBPDatabase } from "idb";

import { wrapOpenDB } from "./wrappers/wrapOpenDB.internal";
import { wrapTransaction } from "./wrappers/wrapTransaction.internal";

import { OpenOptions } from "./OpenOptions";
import {
  NamedSchema,
  Schema,
  StoreKey,
  StoreMap,
  StoreNames,
  StoreValue
} from "./types";

/**
 * Possible modes for transactions created by {@link Database.transaction}.
 */
export const enum TransactionMode {
  ReadOnly = "readonly",
  ReadWrite = "readwrite"
}

/**
 * A wrapper for an IndexedDB connection.
 *
 * Call {@link Database.open} to create a new database connection.
 *
 * @typeparam NamedDBSchema - The schema for the database, along with the
 * allowed names for databases using that schema.
 *
 * @typeparam DBNames - The allowed names for the database. You are unlikely
 * to want to override the default value, derived from `NamedDBSchema`.
 *
 * @typeparam DBSchema - The schema for the database. You are unlikely to want
 * to override the default value, derived from `NamedDBSchema`.
 */
export class Database<
  NamedDBSchema extends NamedSchema<DBNames, DBSchema>,
  // We don't expect to ever override these defaults. They're here to provide
  // convenient type aliases.
  DBNames extends string = NamedDBSchema["dbNames"],
  DBSchema extends Schema = NamedDBSchema["schema"]
> {
  /**
   * Open a new database connection.
   *
   * @typeparam NamedDBSchema - The schema for the database, along with the
   * allowed names for databases using that schema.
   *
   * @typeparam DBNames - The allowed names for the database. You are unlikely
   * to want to override the default value, derived from `NamedDBSchema`.
   *
   * @typeparam DBSchema - The schema for the database. You are unlikely to want
   * to override the default value, derived from `NamedDBSchema`.
   *
   * @param name - The name of the database.
   *
   * @param version - The schema version you want to open. This can only ever
   * increase. See {@link OpenOptions.upgrade} for how to upgrade the schema
   * when changing version.
   *
   * @param openOptions - The options and callbacks for opening the database.
   * See {@link OpenOptions} for details.
   *
   * @returns A promise to an open {@link Database} ready to be accessed.
   */
  static async open<
    NamedDBSchema extends NamedSchema<DBNames, DBSchema>,
    // We don't expect to ever override these defaults. They're here to provide
    // convenient type aliases.
    DBNames extends string = NamedDBSchema["dbNames"],
    DBSchema extends Schema = NamedDBSchema["schema"]
  >(
    name: DBNames,
    version = 1,
    openOptions: OpenOptions<DBSchema> = {}
  ): Promise<Database<NamedDBSchema, DBNames>> {
    const db = await wrapOpenDB(name, version, openOptions);

    return new Database(db);
  }

  /**
   * The name of the IndexedDB database.
   */
  readonly name: string;

  private readonly db: IDBPDatabase<DBSchema>;

  /**
   * Do not use this directly. Use {@link Database.open} to create a new
   * {@link Database}.
   *
   * @ignore
   */
  constructor(db: IDBPDatabase<DBSchema>) {
    this.db = db;

    this.name = this.db.name;
  }

  /**
   * Create or update a value in a store.
   */
  async put<DBStoreName extends StoreNames<DBSchema>>(
    storeName: DBStoreName,
    key: StoreKey<DBSchema, DBStoreName>,
    value: StoreValue<DBSchema, DBStoreName>
  ): Promise<void> {
    await this.db.put(storeName, value, key);
  }

  /**
   * Get a value from a store.
   */
  async get<DBStoreName extends StoreNames<DBSchema>>(
    storeName: DBStoreName,
    key: StoreKey<DBSchema, DBStoreName>
  ): Promise<StoreValue<DBSchema, DBStoreName> | void> {
    const value = await this.db.get(storeName, key);

    if (value !== undefined) {
      return value;
    }
  }

  /**
   * Create a transaction on one or more stores.
   *
   * @param storeNames - The stores to open the transaction over.
   *
   * @param tx - A callback for the transaction to call. Be aware that this
   * callback, although possibly asynchronous, needs to be updating the
   * transaction every tick. If no changes to the transaction happen in a tick,
   * it will automatically commit itself and the transaction will close.
   *
   * @param mode - The mode to open the transaction in.
   */
  async transaction<DBStoreNames extends StoreNames<DBSchema>[]>(
    storeNames: DBStoreNames,
    tx: (stores: StoreMap<DBSchema, DBStoreNames>) => void | Promise<void>,
    mode: TransactionMode = TransactionMode.ReadOnly
  ): Promise<void> {
    const transaction = this.db.transaction(storeNames, mode);

    await wrapTransaction(storeNames, transaction, tx);
  }

  /**
   * Request to close the database connection.
   *
   * If there are no open transactions, this will close immediately. If any
   * transactions are open, it will wait for them to complete before closing.
   * This happens asynchronously, out of band.
   */
  // It would be better if this waited for the connection to close, but
  // `fake-indexeddb` doesn't support the `close` event we need, making it
  // difficult to implement and difficult to test.
  close(): void {
    this.db.close();
  }
}
