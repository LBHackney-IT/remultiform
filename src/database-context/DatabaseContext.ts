import React, { createContext } from "react";

import { Database } from "../database/Database";
import { NamedSchema, Schema } from "../database/types";

/**
 * A wrapper for a React context for passing a database connection to
 * {@link ComponentWrapper|ComponentWrappers}.
 *
 * Normally, you use this by creating an instance somewhere and sharing it
 * between components via {@link DatabaseProvider} and {@link useDatabase}.
 *
 * ```ts
 * type DBSchema = NamedSchema<
 *   "myDatabase" | "myOtherDatabase",
 *   5,
 *   {
 *     favouriteColour: {
 *       key: string;
 *       value: {
 *         r: number;
 *         g: number;
 *         b: number;
 *       };
 *     };
 *   }
 * >;
 *
 * const databasePromise = Database.open("myDatabase", 5);
 * export const DBContext = new DatabaseContext(databasePromise);
 * ```
 */
// This class exists to make it easier to enforce types in other parts of the
// library.
export class DatabaseContext<
  DBSchema extends NamedSchema<string, number, Schema>
> {
  /**
   * An open {@link Database} or a promise that will resolve to an one.
   */
  readonly database: Database<DBSchema> | Promise<Database<DBSchema>>;

  /**
   * The React context itself.
   *
   * @ignore
   */
  readonly context: React.Context<Database<DBSchema> | undefined>;

  /**
   * The consumer component for this context.
   *
   * Use it to wrap components that need the database in the usual React
   * consumer way.
   */
  readonly Consumer: React.ExoticComponent<
    React.ConsumerProps<Database<DBSchema> | undefined>
  >;

  /**
   * Create a new {@link DatabaseContext}.
   *
   * @param openDatabaseOrPromise -  An open {@link Database} or a promise that
   * will resolve to an one. Normally, you would want to pass the return value
   * of {@link Database.open} in here directly.
   */
  constructor(
    openDatabaseOrPromise: Database<DBSchema> | Promise<Database<DBSchema>>
  ) {
    this.database = openDatabaseOrPromise;
    this.context = createContext<Database<DBSchema> | undefined>(undefined);
    this.Consumer = this.context.Consumer;
  }
}
