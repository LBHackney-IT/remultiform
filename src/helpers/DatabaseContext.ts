import React, { createContext } from "react";

import { Database } from "../store/Database";
import { NamedSchema, Schema } from "../store/types";

/**
 * A React context for passing a database connection to
 * {@link PageComponentWrapper|PageComponentWrappers}.
 *
 * ```ts
 * type DBSchema = NamedSchema<
 *   "myDatabase" | "myOtherDatabase",
 *   {
 *     favouriteColour: {
 *       key: string;
 *       value: {
 *         r: number;
 *         g: number;
 *         b: number;
 *       };
 *     };
 *   }
 * >;
 *
 * const DBContext = new DatabaseContext<Database<DBSchema>>();
 * ```
 */
// This class exists to make it easier to enforce types in other parts of the
// library.
export class DatabaseContext<DB extends Database<NamedSchema<string, Schema>>> {
  /**
   * The React context itself.
   *
   * @ignore
   */
  readonly context: React.Context<DB | undefined>;

  constructor() {
    this.context = createContext<DB | undefined>(undefined);
  }
}
