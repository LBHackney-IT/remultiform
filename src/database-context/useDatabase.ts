import React from "react";

import { Database } from "../database/Database";
import { Schema, NamedSchema } from "../database/types";

import { DatabaseContext } from "./DatabaseContext";

/**
 * A React hook to get the {@link Database} instance from the nearest
 * {@link DatabaseProvider} using the same {@link DatabaseContext}.
 *
 * This is only available when using React 16.8 or newer. Use
 * {@link DatabaseContext.Consumer} if you need to use an earlier version.
 */
export const useDatabase = <
  DBSchema extends NamedSchema<string, number, Schema>
>(
  DBContext: DatabaseContext<DBSchema>
): Database<DBSchema> | undefined => {
  return React.useContext(DBContext.context);
};
