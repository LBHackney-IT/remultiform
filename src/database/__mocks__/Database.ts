import { IDBPDatabase } from "idb";

import { promiseToWaitForNextTick } from "../../__tests__/helpers/promise";

import { StoreMap } from "../Store";
import {
  NamedSchema,
  Schema,
  StoreKey,
  StoreNames,
  StoreValue,
} from "../types";

const { Database: ActualDatabase } = jest.requireActual("../Database");

export { ActualDatabase };

export class Database<
  DBSchema extends NamedSchema<string, number, Schema>
> extends ActualDatabase<DBSchema> {
  static async open<DBSchema extends NamedSchema<string, number, Schema>>(
    name: DBSchema["dbNames"],
    version: DBSchema["versions"]
  ): Promise<Database<DBSchema>> {
    await promiseToWaitForNextTick();

    return new Database({ ...jest.fn()(), name, version });
  }

  constructor(db: IDBPDatabase<DBSchema["schema"]>) {
    super(db);
  }

  async put(): Promise<void> {
    await promiseToWaitForNextTick();
  }

  async get<DBStoreName extends StoreNames<DBSchema["schema"]>>(
    storeName: DBStoreName,
    key: StoreKey<DBSchema["schema"], DBStoreName>
  ): Promise<StoreValue<DBSchema["schema"], DBStoreName>> {
    await promiseToWaitForNextTick();

    return `${storeName}/${key}` as StoreValue<DBSchema["schema"], DBStoreName>;
  }

  async transaction<DBStoreNames extends StoreNames<DBSchema["schema"]>[]>(
    _storeNames: DBStoreNames,
    tx: (
      stores: StoreMap<DBSchema["schema"], DBStoreNames>
    ) => void | Promise<void>
  ): Promise<void> {
    await tx(jest.fn()());
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  close(): void {}
}
