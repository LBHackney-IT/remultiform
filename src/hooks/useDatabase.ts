import { useContext } from "react";

import { DatabaseContext } from "../helpers/DatabaseContext";

import { Database } from "../store/Database";
import { Schema, NamedSchema } from "../store/types";

/**
 * A React hook to get the {@link Database} instance from the nearest
 * {@link DatabaseProvider} using the same {@link DatabaseContext}.
 */
export const useDatabase = <DB extends Database<NamedSchema<string, Schema>>>(
  DBContext: DatabaseContext<DB>
): DB | undefined => {
  return useContext(DBContext.context);
};
