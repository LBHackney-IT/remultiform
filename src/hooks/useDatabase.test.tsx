import React from "react";
import { create } from "react-test-renderer";

import { TestErrorBoundary } from "../__fixtures__/components/TestErrorBoundary";

import { spyOnConsoleError } from "../__tests__/helpers/spies";

import { DatabaseContext } from "../helpers/DatabaseContext";

import { useDatabase } from "./useDatabase";
import { Database } from "../store/Database";

jest.mock("../store/Database");

it("returns `undefined` when not a child of a database context provider", () => {
  expect.hasAssertions();

  const DBContext = new DatabaseContext();

  const Tester = (): JSX.Element => {
    const database = useDatabase(DBContext);

    expect(database).toBeUndefined();

    return <div />;
  };

  create(<Tester />);
});

it("returns the database from the nearest database context provider", async () => {
  expect.hasAssertions();

  const DBContext = new DatabaseContext();
  const db = await Database.open("testDBName", 1);

  const Tester = (): JSX.Element => {
    const database = useDatabase(DBContext);

    expect(database).toStrictEqual(db);

    return <div />;
  };

  create(
    <DBContext.context.Provider value={db}>
      <Tester />
    </DBContext.context.Provider>
  );
});

describe("with an unsupported React version", () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let originalReactUseContext: any;

  beforeAll(() => {
    originalReactUseContext = React.useContext;
  });

  afterEach(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (React as any).useContext = originalReactUseContext;
  });

  it("throws if `React.useContext` is undefined", () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    delete (React as any).useContext;

    const DBContext = new DatabaseContext();

    const Tester = (): JSX.Element => {
      useDatabase(DBContext);

      return <div>That was unexpected!</div>;
    };

    const consoleErrorSpy = spyOnConsoleError();

    const component = create(
      <TestErrorBoundary>
        <Tester />
      </TestErrorBoundary>
    );

    expect(consoleErrorSpy.mock.calls).toMatchSnapshot();

    consoleErrorSpy.mockRestore();

    expect(component).toMatchSnapshot();
  });
});
