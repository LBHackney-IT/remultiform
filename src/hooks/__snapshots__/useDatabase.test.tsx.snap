// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with an unsupported React version throws if \`React.useContext\` is undefined 1`] = `
Array [
  Array [
    "The above error occurred in the <Tester> component:
    in Tester
    in TestErrorBoundary

React will try to recreate this component tree from scratch using the error boundary you provided, TestErrorBoundary.",
  ],
]
`;

exports[`with an unsupported React version throws if \`React.useContext\` is undefined 2`] = `
<span>
  Caught error: 
  TypeError: react_1.default.useContext is not a function
</span>
`;
